File: libs/east-py/packages/east-py-cli/east_py_cli/runner.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""IR compilation and execution — uses east-c backend directly.

All formats (JSON, BEAST2, East text) go straight from raw bytes to C
with no Python IR round-trip.
"""

import os
import sys
from collections.abc import Callable, Sequence
from pathlib import Path
from time import perf_counter
from typing import Any

from east.runtime.compiler import compile_from_beast2, compile_from_east, compile_from_json
from east.runtime.errors import EastError
from east.runtime.platform import PlatformFunction
from east.serialization.east_printer import print_east, print_type

from east_py_cli.loader import detect_format, load_platform, load_value, save_value


def _format_size(n: int) -> str:
    if n < 1024:
        return f"{n} B"
    if n < 1024 * 1024:
        return f"{n / 1024:.1f} KB"
    return f"{n / (1024 * 1024):.1f} MB"


def _format_file_size(path: Path) -> str:
    try:
        return _format_size(Path(path).stat().st_size)
    except OSError:
        return "?"


# east-node parity: indexed beast2 collection inputs at or above this many
# bytes open as lazy paged values (EAST_LAZY_INPUT_BYTES overrides; 0
# disables).
_LAZY_INPUT_BYTES_DEFAULT = 64 * 1024 * 1024


def _lazy_input_threshold() -> int:
    env = os.environ.get("EAST_LAZY_INPUT_BYTES", "")
    if env:
        try:
            value = int(env)
        except ValueError:
            value = -1
        if value >= 0:
            return value
    return _LAZY_INPUT_BYTES_DEFAULT


def _load_frozen_input(type_ptr: object, file_path: Path, param_type: Any) -> object:
    """An eagerly-decoded FROZEN input value — task inputs are immutable
    (mutating builtins raise the uniform copy-first error, and frozen
    collections compare by value). beast2 files decode frozen directly;
    other formats load as python values and re-freeze through the beast2
    round-trip, so every format honors the same contract."""
    from east.runtime._compiler_eastc import freeze_value, load_frozen_value

    if Path(file_path).suffix.lower() in (".beast2", ".beast"):
        return load_frozen_value(type_ptr, Path(file_path).read_bytes())
    return freeze_value(type_ptr, load_value(file_path, param_type))


def _compile_ir_file(ir_file: Path, platform_fns: list[PlatformFunction]) -> tuple[Callable, bool]:
    """Compile an IR file straight from its bytes — no Python IR round-trip,
    a single file read — returning the compiled function and whether it is
    async."""
    fmt = detect_format(ir_file)
    is_async = False
    if fmt == "json":
        data = ir_file.read_bytes()
        is_async = b'"AsyncFunction"' in data[:200]
        return compile_from_json(data, platform_fns, is_async), is_async
    if fmt == "beast2":
        return compile_from_beast2(ir_file.read_bytes(), platform_fns, is_async), is_async
    if fmt == "east":
        text = ir_file.read_text(encoding="utf-8")
        is_async = "AsyncFunction" in text[:200]
        return compile_from_east(text, platform_fns, is_async), is_async
    raise ValueError(f"Unknown IR format: {fmt}")


def print_result(timings: dict[str, float], peak_bytes: int) -> None:
    """Print a unit's result as ``-v`` shows it: where the time went, in
    milliseconds, and the process's peak memory."""
    print("\nTiming:", file=sys.stderr)
    print(f"  Load:     {timings['load']:8.1f} ms", file=sys.stderr)
    print(f"  Compile:  {timings['compile']:8.1f} ms", file=sys.stderr)
    print(f"  Execute:  {timings['execute']:8.1f} ms", file=sys.stderr)
    print(f"  Output:   {timings['output']:8.1f} ms", file=sys.stderr)
    print(f"  Total:    {sum(timings.values()):8.1f} ms", file=sys.stderr)
    print("\nMemory:", file=sys.stderr)
    if peak_bytes >= 1024 * 1024:
        print(f"  Peak RSS: {peak_bytes / (1024 * 1024):8.1f} MB", file=sys.stderr)
    else:
        print(f"  Peak RSS: {peak_bytes / 1024:8.0f} KB", file=sys.stderr)


def _open_inputs(handle: Any, input_files: Sequence[Path], lazy: Callable[[int, int], bool],
                 verbose: bool) -> tuple[list[object], list[int]]:
    """A program's inputs, always FROZEN — task inputs are immutable; mutating
    one raises the uniform copy-first error — and which of them opened lazily.

    A beast2 collection input ``i`` of ``size`` bytes for which ``lazy(i,
    size)`` holds opens as a lazy paged value (segment-fed iteration and keyed
    reads at O(segment) decoded memory — #505), and because frozen collapses
    the shape gate, nested-container element shapes open lazily too. A lazily
    opened file is MAPPED, never read whole: the paged value serves its reads
    from the mapping, so the input's residency is the page cache and the heap
    holds one decoded segment at a time. Anything not pageable falls back to
    the whole (frozen) decode, exactly like east-node's runner. A file
    holding a manifest is the collection it names, its segments in the
    directory beside it, and it weighs its segments rather than its own few
    kilobytes.
    """
    from east.runtime._compiler_eastc import (
        load_frozen_manifest,
        manifest_segment_bytes,
        open_manifest_file,
        open_paged_file,
    )

    input_types = handle.get_input_types()
    inputs: list[object] = []
    lazy_inputs: list[int] = []
    for i, (file_path, param_type) in enumerate(zip(input_files, input_types, strict=False)):
        beast2 = Path(file_path).suffix.lower() in (".beast2", ".beast")
        collection = getattr(param_type, "type", None) in ("Array", "Set", "Dict")
        weight = manifest_segment_bytes(file_path) if beast2 and collection else None
        size = Path(file_path).stat().st_size + (weight or 0)
        opened = None
        if beast2 and collection and lazy(i, size):
            try:
                opener = open_paged_file if weight is None else open_manifest_file
                opened = opener(handle._input_types[i], file_path, frozen=True)
            except (OSError, ValueError):
                # Not a container of the parameter's type, or a file that
                # cannot be mapped: the whole decode below reads it and
                # reports the real error.
                opened = None
        if opened is not None:
            lazy_inputs.append(i)
            if verbose:
                print(f"  input {i}: opened lazily — mapped from the file", file=sys.stderr)
            inputs.append(opened)
        elif weight is not None:
            inputs.append(load_frozen_manifest(handle._input_types[i], file_path))
        else:
            inputs.append(_load_frozen_input(handle._input_types[i], file_path, param_type))
    return inputs, lazy_inputs


def execute_unit(unit_path: Path) -> dict[str, Any]:
    """Execute a unit — ``east-py exec``: do its work, write its output, and
    record its result where the unit says.

    The unit, the result and the output writers are east-c's (east/unit.h),
    the very code the east-c CLI runs, so the two runners read the same units
    and write the same bytes; east-node implements the same protocol, and the
    conformance corpus holds the three to it.

    Returns the result: ``ok``, a failure's ``message`` and ``locations``
    (``(filename, line, column)``, innermost first), ``peak_bytes`` and
    ``timings``. A failure is the result's, never a raise.

    Raises:
        ValueError: If the file does not hold a unit.
        OSError: If the result cannot be written.
    """
    from east.serialization._beast2_eastc import (
        _peak_bytes,
        _read_unit,
        _set_thread_limit,
        _write_unit_result,
    )

    unit = _read_unit(unit_path)
    # The grant caps every pool east-c starts; one thread frames every output
    # inline.
    _set_thread_limit(min(max(unit["threads"], 1), 1024))
    timings = {"load": 0.0, "compile": 0.0, "execute": 0.0, "output": 0.0}
    mark = perf_counter()

    def lap(phase: str) -> None:
        nonlocal mark
        now = perf_counter()
        timings[phase] += (now - mark) * 1000
        mark = now

    result: dict[str, Any]
    try:
        platform_fns: list[PlatformFunction] = []
        for package in unit["platforms"]:
            platform_fns.extend(load_platform(package))
        if unit["merge"]:
            _merge_work(unit, platform_fns, lap)
        else:
            _run_work(unit, platform_fns, lap)
        result = {"ok": True, "message": None, "locations": []}
    except EastError as e:
        result = {
            "ok": False,
            "message": e.message,
            "locations": [(frame["filename"], frame["line"], frame["column"]) for frame in e.location],
        }
    except Exception as e:  # noqa: BLE001 - every failure is the result's
        result = {"ok": False, "message": str(e), "locations": []}
    result["peak_bytes"] = _peak_bytes()
    result["timings"] = timings
    _write_unit_result(unit["result"], result["ok"], result["message"], result["locations"],
                       result["peak_bytes"], timings)
    return result


def _emit_parameter(kind: str, emit_type: Any) -> list[Any]:
    """The emit parameter's argument types, checked against the output's kind:
    one for an array, a set or a fold, a key and a value for a dict."""
    arity = 2 if kind == "dict" else 1
    if getattr(emit_type, "type", None) != "Function" or len(emit_type.value["inputs"]) != arity:
        raise ValueError(
            f"exec: a {kind} output is emitted: the program's trailing parameter must be emit, "
            f"a function of {arity} argument{'' if arity == 1 else 's'}, got "
            f"{print_type(emit_type)}")
    return list(emit_type.value["inputs"])


def _run_work(unit: dict[str, Any], platform_fns: list[PlatformFunction],
              lap: Callable[[str], None]) -> None:
    """A run unit: the program evaluated on its inputs, its output written by
    kind."""
    from east.runtime._compiler_eastc import _eastc_call, load_frozen_value
    from east.serialization._beast2_eastc import _UnitSinkCore, _write_unit_value

    compiled, _ = _compile_ir_file(Path(unit["program"]), platform_fns)
    handle = compiled._eastc_handle  # type: ignore[attr-defined]
    input_types = handle.get_input_types()
    output = unit["output"]
    # Every kind but a value is emitted, through the trailing parameter.
    emitted = output["kind"] != "value"
    if emitted and not input_types:
        raise ValueError(
            f"exec: a {output['kind']} output is emitted: the program's trailing parameter "
            "must be emit, a function")
    params = input_types[:-1] if emitted else input_types
    if len(unit["inputs"]) != len(params):
        sig_params = ", ".join(print_type(t) for t in input_types)
        raise ValueError(
            f"Function expects {len(params)} inputs, got {len(unit['inputs'])}\n"
            f"Signature: ({sig_params}) -> {print_type(handle.get_output_type())}")

    sink = None
    if emitted:
        emit_types = _emit_parameter(output["kind"], input_types[-1])
        merge = combine = zero = None
        if output["kind"] == "dict" and output["merge"] is not None:
            merge = _compile_ir_file(Path(output["merge"]), platform_fns)[0]
        if output["kind"] == "fold":
            combine = _compile_ir_file(Path(output["combine"]), platform_fns)[0]
            zero = load_frozen_value(emit_types[0], Path(output["zero"]).read_bytes())
        sink = _UnitSinkCore(output["kind"], emit_types, output["path"], merge=merge,
                             combine=combine, zero=zero)
    lap("compile")

    threshold = _lazy_input_threshold()
    inputs, _ = _open_inputs(handle, [Path(p) for p in unit["inputs"]],
                             lambda _i, size: threshold > 0 and size >= threshold, False)
    if sink is not None:
        inputs.append(sink.function_value())
    lap("load")

    returned = _eastc_call(handle._compiled, handle._input_types, handle._output_type, tuple(inputs))
    lap("execute")
    if sink is not None:
        sink.finish()
    else:
        _write_unit_value(handle.get_output_type(), output["path"], returned)
    lap("output")


def _merge_work(unit: dict[str, Any], platform_fns: list[PlatformFunction],
                lap: Callable[[str], None]) -> None:
    """A merge unit: set or dict parts merged into one run, or fold partials
    folded in order, starting at zero."""
    from east.runtime._compiler_eastc import load_frozen_value
    from east.serialization._beast2_eastc import _unit_merge_runs, _write_unit_value

    output = unit["output"]
    kind = output["kind"]
    if kind == "array":
        raise ValueError(
            "exec: an array's parts are concatenated, never merged: a merge unit takes set, "
            "dict or fold parts")
    if kind == "value":
        raise ValueError("exec: a value has no parts: a merge unit takes set, dict or fold parts")
    if kind in ("set", "dict"):
        merge = None
        if kind == "dict" and output["merge"] is not None:
            merge = _compile_ir_file(Path(output["merge"]), platform_fns)[0]
        lap("load")
        _unit_merge_runs(unit["inputs"], output["path"], kind, unit["range"], merge)
        lap("execute")
        return
    # A merge of fold partials has no program to say what was emitted: the
    # combine's own type does.
    combine = _compile_ir_file(Path(output["combine"]), platform_fns)[0]
    ins = combine._eastc_handle.get_input_types()  # type: ignore[attr-defined]
    value_type = combine._eastc_handle.get_output_type()  # type: ignore[attr-defined]
    if len(ins) != 2 or ins[0] != value_type or ins[1] != value_type:
        raise ValueError(
            f"exec: combine: expected a function (T, T) -> T (T = {print_type(value_type)}), "
            f"got ({', '.join(print_type(t) for t in ins)}) -> {print_type(value_type)}")
    acc = load_frozen_value(value_type, Path(output["zero"]).read_bytes())
    lap("compile")
    for part in unit["inputs"]:
        acc = combine(acc, load_frozen_value(value_type, Path(part).read_bytes()))
    lap("execute")
    _write_unit_value(value_type, output["path"], acc)
    lap("output")


def run_program(
    ir_file: Path,
    platform_fns: list[PlatformFunction],
    packages: list[str],
    input_files: list[Path],
    output_file: Path | None = None,
    verbose: bool = False,
) -> object:
    """Run an East IR program: its result written to ``output_file`` in the
    format its extension names, or printed as East text."""
    t0 = perf_counter()

    # Compile directly from raw data — no Python IR round-trip, single file read
    compiled, is_async = _compile_ir_file(ir_file, platform_fns)

    t1 = perf_counter()

    handle = compiled._eastc_handle  # type: ignore[attr-defined]
    input_types = handle.get_input_types()
    output_type = handle.get_output_type()

    # Validate input count
    if len(input_files) != len(input_types):
        sig_params = ", ".join(print_type(t) for t in input_types)
        raise ValueError(
            f"Function expects {len(input_types)} inputs, got {len(input_files)}\n"
            f"Signature: ({sig_params}) -> {print_type(output_type)}"
        )

    # Verbose header
    if verbose:
        print(f"Running: {ir_file}  ({_format_file_size(ir_file)})", file=sys.stderr)

        if packages:
            print(
                f"Platform: {len(packages)} package(s), {len(platform_fns)} function(s)",
                file=sys.stderr,
            )
            for p in packages:
                print(f"  - {p}", file=sys.stderr)

        print(
            f"Function: {len(input_types)} inputs, {'async' if is_async else 'sync'}",
            file=sys.stderr,
        )
        for i, (file_path, param_type) in enumerate(zip(input_files, input_types, strict=False)):
            print(f"  input {i}: {file_path}  ({_format_file_size(file_path)})", file=sys.stderr)
            print(f"    {print_type(param_type)}", file=sys.stderr)
        print("  return:", file=sys.stderr)
        print(f"    {print_type(output_type)}", file=sys.stderr)

    # Indexed beast2 collection inputs open lazily at or above the size
    # threshold.
    threshold = _lazy_input_threshold()
    inputs, lazy_inputs = _open_inputs(
        handle, input_files, lambda _i, size: threshold > 0 and size >= threshold, verbose)

    t2 = perf_counter()

    # Execute via east-c
    from east.runtime._compiler_eastc import _eastc_call

    result = _eastc_call(handle._compiled, handle._input_types, handle._output_type, tuple(inputs))

    t3 = perf_counter()

    # Output
    if output_file is not None:
        save_value(output_file, result, output_type)
        if verbose:
            print(f"Output: {output_file}  ({_format_file_size(output_file)})", file=sys.stderr)
            print(f"  {print_type(output_type)}", file=sys.stderr)
    else:
        print(print_east(result, output_type))

    t4 = perf_counter()

    if verbose:
        from east.serialization._beast2_eastc import _peak_bytes

        print_result({
            "load": (t1 - t0) * 1000,
            "compile": (t2 - t1) * 1000,
            "execute": (t3 - t2) * 1000,
            "output": (t4 - t3) * 1000,
        }, _peak_bytes())

        # What each lazy input's reads came to — the account residency
        # cannot give on a mapping, where the kernel decides how much of a
        # touched file is resident.
        from east.runtime._compiler_eastc import paged_value_stats

        for i in lazy_inputs:
            stats = paged_value_stats(getattr(inputs[i], "_east_c_paged", 0))
            if stats is None:
                continue
            segments, decoded, fences, hydrated = stats
            if hydrated:
                print(f"  input {i}: decoded whole (an operation the pager cannot serve)",
                      file=sys.stderr)
            else:
                print(f"  input {i}: {decoded} of {segments} segments decoded, {fences} fences probed",
                      file=sys.stderr)

    return result

File: libs/east-node/packages/east-node-cli/src/runner.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

import { readFileSync, statSync, writeFileSync } from 'fs';
import { extname } from 'path';
import {
    EastIR,
    encodeBeast2For,
    encodeBeast2PagedFor,
    encodeEastFor,
    encodeJSONFor,
    isTypeValueEqual,
    printFor,
    variant,
    type UnitResult,
} from '@elaraai/east';
import type { PlatformFunction, EastTypeValue } from '@elaraai/east/internal';
import { printTypeValue } from '@elaraai/east/internal';
import { inputBytes, lazyInputBytesRead, loadEastIR, loadInput, loadInputLazy } from './loader.js';

function now(): bigint { return process.hrtime.bigint(); }
function elapsed(start: bigint, end: bigint): number { return Number(end - start) / 1e6; }

function formatSize(bytes: number): string {
    if (bytes < 1024) return `${bytes} B`;
    if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
    return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

/** A file's size for the verbose account, or `?` when it cannot be read. @internal */
export function formatFileSize(path: string): string {
    try { return formatSize(statSync(path).size); } catch { return '?'; }
}

/**
 * A refusal of the command line itself — a function the inputs given do not
 * fit.
 *
 * These are the user's errors, not the program's, and east-c and east-py
 * answer them with one sentence on stderr. Marking them lets the CLI do the
 * same instead of printing the JS stack it prints for an error thrown from
 * East or from a platform function, where the frames are the diagnosis.
 */
export class UsageError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'UsageError';
    }
}

/** Default size threshold above which collection inputs open lazily. */
const LAZY_INPUT_BYTES_DEFAULT = 64 * 1024 * 1024;

/** Resolves the lazy-open threshold: the environment, else the default. An
 *  unset or empty `EAST_LAZY_INPUT_BYTES` falls through to the 64 MiB default
 *  (`Number('')` is `0`, which would silently DISABLE lazy opening); invalid
 *  or negative values fall through too, matching east-c and east-py. Exported
 *  for `exec` and the spec. @internal */
export function lazyThreshold(): number {
    const raw = process.env.EAST_LAZY_INPUT_BYTES;
    if (raw !== undefined && raw !== '') {
        const env = Number(raw);
        if (Number.isFinite(env) && env >= 0) return env;
    }
    return LAZY_INPUT_BYTES_DEFAULT;
}

/**
 * Runs an East IR program.
 */
export async function runProgram(
    irPath: string,
    platformFns: PlatformFunction[],
    packages: string[],
    inputPaths: string[],
    outputPath?: string,
    verbose = false,
): Promise<unknown> {
    const t0 = now();

    // Load as an EastIR bundle so source_map travels with the IR and error
    // frames resolve end-to-end.
    const eastIR = loadEastIR(irPath);
    const ir = eastIR.ir;
    const isAsync = eastIR instanceof EastIR ? false : true;

    // Get the function's input/output types from the IR.
    const inputTypes = (ir as any)?.value?.type?.value?.inputs ?? [];
    const outputType = ((ir as any)?.value?.type?.value?.output ?? null) as EastTypeValue | null;

    if (inputPaths.length !== inputTypes.length) {
        // east-c / east-py parity, down to the signature line: one sentence
        // for one condition, whichever runner the task declares.
        const signature = `(${(inputTypes as EastTypeValue[]).map((t) => printTypeValue(t)).join(', ')}) -> ` +
            `${outputType !== null ? printTypeValue(outputType) : '?'}`;
        throw new UsageError(
            `Function expects ${inputTypes.length} inputs, got ${inputPaths.length}\nSignature: ${signature}`,
        );
    }

    // Verbose header
    if (verbose) {
        console.error(`Running: ${irPath}  (${formatFileSize(irPath)})`);

        if (packages.length > 0) {
            console.error(`Platform: ${packages.length} package(s), ${platformFns.length} function(s)`);
            for (const p of packages) console.error(`  - ${p}`);
        }

        console.error(`Function: ${inputTypes.length} inputs, ${isAsync ? 'async' : 'sync'}`);
        for (let i = 0; i < inputPaths.length; i++) {
            const t = printTypeValue(inputTypes[i]!);
            console.error(`  input ${i}: ${inputPaths[i]}  (${formatFileSize(inputPaths[i]!)})`);
            console.error(`    ${t}`);
        }
        if (outputType) {
            console.error(`  return:`);
            console.error(`    ${printTypeValue(outputType)}`);
        }
    }

    // Load inputs — always frozen (task inputs are immutable; mutating one
    // throws the uniform copy-first error). Beast2 collection inputs open
    // lazily at or above the size threshold, so a sparse read into a huge
    // indexed input stops paying a whole decode — and because frozen collapses
    // the shape gate, nested-container element shapes open lazily too. The
    // size is the value's: an input staged as a manifest is a small file
    // naming large ones.
    const threshold = lazyThreshold();
    const inputs: unknown[] = [];
    const lazyInputs: number[] = [];
    for (let i = 0; i < inputPaths.length; i++) {
        const lazy = threshold > 0 && inputBytes(inputPaths[i]!) >= threshold ? loadInputLazy(inputPaths[i]!) : undefined;
        if (lazy !== undefined) {
            lazyInputs.push(i);
            if (verbose) console.error(`  input ${i}: opened lazily — paged from the file`);
        }
        inputs.push(lazy !== undefined ? lazy : loadInput(inputPaths[i]!, inputTypes[i]!));
    }
    /** The verbose summary's account of each lazy input: what paging came
     *  to, against the size of the value. */
    const reportLazyReads = (): void => {
        for (const i of lazyInputs) {
            const read = lazyInputBytesRead(inputs[i]);
            if (read !== undefined) console.error(`  input ${i}: ${formatSize(read)} read of ${formatSize(inputBytes(inputPaths[i]!))}`);
        }
    };

    const t1 = now();

    let result: unknown;
    if (!isAsync) {
        const compiled = (eastIR as EastIR<any, any>).compile(platformFns);
        const t2 = now();

        result = compiled(...inputs);
        const t3 = now();

        const t4 = maybeWriteOutput(outputPath, result, outputType, verbose);

        if (verbose) {
            printResult({
                outcome: variant('ok', null),
                peakBytes: peakBytes(),
                timings: { load: elapsed(t0, t1), compile: elapsed(t1, t2), execute: elapsed(t2, t3), output: elapsed(t3, t4) },
            });
            reportLazyReads();
        }
    } else {
        const compiled = (eastIR as any).compile(platformFns);
        const t2 = now();

        result = await compiled(...inputs);
        const t3 = now();

        const t4 = maybeWriteOutput(outputPath, result, outputType, verbose);

        if (verbose) {
            printResult({
                outcome: variant('ok', null),
                peakBytes: peakBytes(),
                timings: { load: elapsed(t0, t1), compile: elapsed(t1, t2), execute: elapsed(t2, t3), output: elapsed(t3, t4) },
            });
            reportLazyReads();
        }
    }

    return outputPath ? undefined : result;
}

/** A compiled `(K, V, V) -> V` fold of equal keys. */
export type MergeFunction = (key: unknown, acc: unknown, value: unknown) => unknown;

/**
 * Loads a merge function: an IR file holding a `(K, V, V) -> V` East function
 * over the given key and value types, compiled with the unit's platforms.
 * Shared by `exec`'s dict output and the blob merge.
 *
 * @param path - the IR file (any format the IR positional accepts)
 * @param keyType - the key type
 * @param valueType - the value type
 * @param platformFns - the unit's platforms
 * @param subject - what the function must match, for the message: `the emit
 *   parameter` or `the inputs`
 * @returns the compiled merge function
 * @throws {Error} When the IR is not a function of that shape, naming the
 *   expected and the actual types.
 */
export function loadMergeFunction(path: string, keyType: EastTypeValue, valueType: EastTypeValue, platformFns: PlatformFunction[], subject: string): MergeFunction {
    const bundle = loadEastIR(path);
    const fnType = (bundle.ir as any).value.type as EastTypeValue;
    const shape = fnType.type === 'Function' ? fnType.value as { inputs: EastTypeValue[]; output: EastTypeValue } : null;
    if (shape === null || shape.inputs.length !== 3 ||
        !isTypeValueEqual(shape.inputs[0]!, keyType) ||
        !isTypeValueEqual(shape.inputs[1]!, valueType) ||
        !isTypeValueEqual(shape.inputs[2]!, valueType) ||
        !isTypeValueEqual(shape.output, valueType)) {
        throw new Error(
            `merge function: expected a function (K, V, V) -> V matching ${subject} ` +
            `(K = ${printTypeValue(keyType)}, V = ${printTypeValue(valueType)}), got ${printTypeValue(fnType)}`,
        );
    }
    return (bundle as EastIR<any, any>).compile(platformFns) as MergeFunction;
}

function maybeWriteOutput(outputPath: string | undefined, result: unknown, outputType: EastTypeValue | null, verbose: boolean): bigint {
    if (!outputPath) {
        // Print to stdout as .east format (matches east-c / east-py)
        if (outputType) {
            const printer = printFor(outputType as any);
            console.log(printer(result));
        }
        return now();
    }
    if (!outputType) return now();
    writeOutput(outputPath, result, outputType);
    const t = now();
    if (verbose) {
        console.error(`Output: ${outputPath}  (${formatFileSize(outputPath)})`);
        console.error(`  ${printTypeValue(outputType)}`);
    }
    return t;
}

/**
 * Prints a unit's result as `-v` shows it: where the time went, and the
 * process's peak memory.
 *
 * @param result - the result
 */
export function printResult(result: UnitResult): void {
    const { load, compile, execute, output } = result.timings;
    console.error('\nTiming:');
    console.error(`  Load:     ${load.toFixed(1).padStart(8)} ms`);
    console.error(`  Compile:  ${compile.toFixed(1).padStart(8)} ms`);
    console.error(`  Execute:  ${execute.toFixed(1).padStart(8)} ms`);
    console.error(`  Output:   ${output.toFixed(1).padStart(8)} ms`);
    console.error(`  Total:    ${(load + compile + execute + output).toFixed(1).padStart(8)} ms`);
    console.error('\nMemory:');
    console.error(`  Peak RSS: ${(Number(result.peakBytes) / (1024 * 1024)).toFixed(1).padStart(8)} MB`);
}

/**
 * The process's peak resident memory in bytes: VmHWM on Linux, where
 * `ru_maxrss` carries a parent's peak across exec, and `ru_maxrss` elsewhere.
 *
 * @returns the peak
 */
export function peakBytes(): bigint {
    try {
        const hwm = /^VmHWM:\s+(\d+) kB$/m.exec(readFileSync('/proc/self/status', 'utf8'));
        if (hwm !== null) return BigInt(hwm[1]!) * 1024n;
    } catch {
        // No /proc: not Linux.
    }
    return BigInt(process.resourceUsage().maxRSS) * 1024n;
}

/** Whether an output root type is a collection (Array/Set/Dict). */
function isCollectionRoot(type: EastTypeValue): boolean {
    return type.type === 'Array' || type.type === 'Set' || type.type === 'Dict';
}

function writeOutput(filePath: string, value: unknown, type: unknown): void {
    const ext = extname(filePath).toLowerCase();
    switch (ext) {
        case '.beast2':
        case '.beast': {
            // Collection-rooted outputs are ALWAYS segmented + indexed, cut by
            // the content-defined rule, so e3's paged dataset reads can seek —
            // one encoding per logical value, at every size.
            const encoder = isCollectionRoot(type as EastTypeValue)
                ? encodeBeast2PagedFor(type as any)
                : encodeBeast2For(type as any);
            writeFileSync(filePath, encoder(value));
            break;
        }
        case '.east': {
            const encoder = encodeEastFor(type as any);
            writeFileSync(filePath, encoder(value));
            break;
        }
        case '.json': {
            const encoder = encodeJSONFor(type as any);
            writeFileSync(filePath, encoder(value));
            break;
        }
        default:
            throw new Error(
                `Unsupported output file extension "${ext}". ` +
                `Supported extensions: .beast2, .beast, .east, .json`,
            );
    }
}
